File: AoC2015/Day1/aoc2015_day1_part_b.cpp
/**
 * Now, given the same instructions, find the position 
 * of the first character that causes him to enter the 
 * basement (floor -1).
 */
#include<iostream> // IO processing from stdin/stdout
#include<string> // String class for CPP
#include <cstdlib> // Absolute value function
#include <fstream> // File handling for IO file streams
using namespace std;

/** Function Prototypes */
int findBasementPosition(string instructions);

/** Driver Program */
int main() {
    // Declare file stream for file handling
    fstream file;
    // Declare string to hold Santa's instructions
    string instructions;
    // Open file for reading
    file.open("aoc2015_day1_input.txt", ios::in);
    // Process the string in the file
    file >> instructions;
    
    /** Assertions */
    // Some assertions to check the function is working correctly
    // Sample I/O assertions
    cout << boolalpha; // Set flag to convert 1 and 0 to true and false
    cout << "TEST1: " << (findBasementPosition(")") == 1) << "\n";
    cout << "TEST2: " << (findBasementPosition("()())") == 5) << "\n";
    // Final assertion on challenge input
    cout << "TESTN: " << (findBasementPosition(instructions) == 1783) << "\n";
}

/** Function Definitions */
int findBasementPosition(string instructions) {
    /**
     * Find position of the first character that causes 
     * santa to enter the basement floor (-1).
     */
    int floorNumber = 0, instructionPosition = 1;
    // Iterate over instructions
    for (char instruction : instructions) {
        if (instruction == '(') {
            floorNumber++;
        } else if (instruction == ')') {
            floorNumber--;
            if (floorNumber < 0) return instructionPosition;
        }
        instructionPosition++;
    }
}

File: AoC2015/Day1/aoc2015_day1_part_b.py
"""
Now, given the same instructions, find the position 
of the first character that causes him to enter the 
basement (floor -1).
"""
def find_basement_position(instructions):
    """
    Find position of the first character that causes 
    santa to enter the basement floor (-1).
    """
    # Santa starts at floor zero
    # Instruction position starts at one
    floor_number, instruction_position = 0, 1
    # Iterate over instructions
    for instruction in instructions:
        if instruction == '(':
            floor_number += 1
        elif instruction == ')':
            floor_number -= 1
            if floor_number < 0:
                return instruction_position
        instruction_position += 1

if __name__ == '__main__':
    # This variable holds the instructions for Santa
    instructions = None
    # Open the file for processing
    with open("aoc2015_day1_input.txt", 'r') as f:
        # Reads contents of file as a single string
        instructions = f.read()
        # Close the file after processing
        f.close()
    # Some assertions to check the function is working correctly
    assert find_basement_position(")") == 1
    assert find_basement_position("()())") == 5
    # Print the solution to the challenge with the text file input
    print(find_basement_position(instructions))